data/beaconrestapi/src/main/java/tech/pegasys/artemis/beaconrestapi/beaconhandlers/BeaconStateHandler.java

```diff
@@ -16,40 +16,52 @@
 import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
 import static javax.servlet.http.HttpServletResponse.SC_NOT_FOUND;
 
+import com.google.common.primitives.UnsignedLong;
 import io.javalin.http.Context;
 import io.javalin.http.Handler;
 import io.javalin.plugin.openapi.annotations.HttpMethod;
 import io.javalin.plugin.openapi.annotations.OpenApi;
 import io.javalin.plugin.openapi.annotations.OpenApiContent;
 import io.javalin.plugin.openapi.annotations.OpenApiParam;
 import io.javalin.plugin.openapi.annotations.OpenApiResponse;
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.Logger;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import org.apache.commons.lang3.StringUtils;
 import org.apache.tuweni.bytes.Bytes32;
 import tech.pegasys.artemis.beaconrestapi.schema.BadRequest;
 import tech.pegasys.artemis.datastructures.state.BeaconState;
 import tech.pegasys.artemis.provider.JsonProvider;
-import tech.pegasys.artemis.storage.ChainStorageClient;
-import tech.pegasys.artemis.storage.Store;
+import tech.pegasys.artemis.storage.CombinedChainDataClient;
+import tech.pegasys.artemis.util.async.SafeFuture;
 
 public class BeaconStateHandler implements Handler {
   public static final String ROUTE = "/beacon/state";
-  private final Logger LOG = LogManager.getLogger();
-  private final ChainStorageClient client;
+  public static final String ROOT = "root";
+  public static final String SLOT = "slot";
+  private final CombinedChainDataClient combinedClient;
   private final JsonProvider jsonProvider;
 
-  public BeaconStateHandler(ChainStorageClient client, JsonProvider jsonProvider) {
-    this.client = client;
+  public BeaconStateHandler(
+      final CombinedChainDataClient combinedClient, final JsonProvider jsonProvider) {
+    this.combinedClient = combinedClient;
     this.jsonProvider = jsonProvider;
   }
 
-  private BeaconState queryByRootHash(String root) {
-    Bytes32 root32 = Bytes32.fromHexString(root);
-    Store store = client.getStore();
-    if (store == null) {
-      return client.getBlockState(root32).orElse(null);
-    }
-    return store.getBlockState(root32);
+  private Optional<BeaconState> queryByRootHash(final String root)
+      throws ExecutionException, InterruptedException {
+    final Bytes32 root32 = Bytes32.fromHexString(root);
+    final SafeFuture<Optional<BeaconState>> future = combinedClient.getStateAtBlock(root32);
+    return future.get();
```

Comment:
It looks like Javalin's `ctx.result` can be passed a future and it will handle it all for us rather than blocking the thread.  So this and `queryBySlot` would return a `SafeFuture` and use `thenApply` to do the processing async.

Reply:
good plan